### DISCUSSION TITLE
ContainEquivalentOf for a collection

### DISCUSSION THREAD

**Discussion by comecme** [fluentassertions/fluentassertions#1988] [Ideas]
I'm looking for a way to assert that a collection contains a set of objects. I want the behavior of BeEquivalentTo combined with Contain.

So let's say my orderDtos variable contains 4 items. I want to assert that two specific items exist in the collection. Not the exact same instance of the OrderDto, but an equivalent.

This won't work. This asserts that orderDtos contains a single item that is equivalent to an array with the two items.
```c#
orderDtos.Should().ContainEquivalentOf(new[] {
  new OrderDto(1),
  new OrderDto(2)
});
``` 

This fails because orderDtos contains more than the two orderDto's.
```c#
orderDtos.Should().BeEquivalentTo(
  new OrderDto(1),
  new OrderDto(2));
```
This works but seems complicated:
```c#
orderDtos.Should().ContainEquivalentOf(
  new OrderDto(1).And.ContainEquivalentOf(
  new OrderDto(2));
```

For simple types, the Contain method does what I want:
```c#
orderDtos.Select(o => o.Id).Should().Contain(new[] { 1, 2 });
```

But as I said, I'd like to compare the object. I would even want to be able to compare with an anonymous type.

Is there a method that is able to do this that I am overlooking? Or is this just not possible?

**Comment by jnyrup** [comment]
There's nothing built-in to do this, but a `ContainEquivalentsOf<TExpectation>(IEnumerable<TExpectation> expectations)` would be a nice addition.

For the general case a chain of calls to `ContainEquivalentOf` will not work as they may match the same element multiple times.
```cs
class MyDto
{
    public int Foo { get; set; }

    public int Bar { get; set; }
}

[Fact]
public void False_positive()
{
    var dtos = new[]
    {
        new MyDto { Foo = 13, Bar = 37 }
    };

    dtos.Should().ContainEquivalentOf(new { Foo = 13 })
        .And.ContainEquivalentOf(new { Bar = 37 });
}
```

I added some related notes about how to use [MaximumMatchingProblem](https://github.com/fluentassertions/fluentassertions/blob/develop/Src/FluentAssertions/Collections/MaximumMatching/MaximumMatchingProblem.cs) to solve this in #1486.

**Comment by inikulshin** [comment]
`ContainEquivalentOfInOrder<TExpectation>(IEnumerable<TExpectation> expectations)` would be a nice addition too.
And this one is much easier to implement.
